src/libraries/System.ComponentModel.Annotations/src/System/ComponentModel/DataAnnotations/ValidationAttributeStore.cs

```diff
@@ -230,8 +232,60 @@ internal bool TryGetPropertyStoreItem(string propertyName, [NotNullWhen(true)] o
                     propertyStoreItems[property.Name] = item;
                 }
 
+                return CombineMetadataTypeAttributes(propertyStoreItems);
+            }
+
+            private Dictionary<string, PropertyStoreItem> CombineMetadataTypeAttributes(Dictionary<string, PropertyStoreItem> propertyStoreItems)
+            {
+                // check if metadata attribute is present on the type
+                MetadataTypeAttribute? attribute = ((MetadataTypeAttribute[])Attribute.GetCustomAttributes(_type, typeof(MetadataTypeAttribute))).FirstOrDefault();
```

Comment:
I'm not sure, but shouldn't this go through AssociatedMetadataTypeTypeDescriptor? Even if not, it should probably go through TypeDescriptorCache.GetAssociatedMetadataType.

Reply:
On .NET Framework - `CreatePropertyStoreItems()` called `TypeDescriptor.GetProperties`, and then filtered the attributes that were explicitly specified. See

https://referencesource.microsoft.com/#System.ComponentModel.DataAnnotations/DataAnnotations/ValidationAttributeStore.cs,211

Reply:
Yes copying the .NET Framework code over works as long as you also use 
`TypeDescriptor.AddProviderTransparent( new AssociatedMetadataTypeTypeDescriptionProvider(typeof(Person), typeof(PersonMetadata)), typeof(Person));`

There was commentary in the code and elsewhere in forums (I can't remember where) indicating that the TypeDescriptor approach was not the way forward, hence this PR. 

Thanks again to @eerhardt for his help.

Reply:
You may be referring to https://github.com/dotnet/runtime/issues/46678#issuecomment-787738854. However, as @bricelam and I wrote, we're looking to align the .NET Core implementation here with the .NET Framework, by copying the code across from the .NET Framework reference sources, rather than writing new support here as you've done (if I understand everything correctly, that corresponds to solution 1 in the linked comment).

Reply:
Shall I copy that over? Or are there other things you want to check/consider first?

Reply:
I think copying that over and thereby aligning the .NET Core implementation to the .NET Framework one is what we're looking for - I don't think there's something else to check.

Reply:
The build is failing due to trim analysis errors. I've tried for a while but I can't figure out how to suppress them. Any pointers appreciated.

Reply:
The annotation on the `_type` field doesn't match the required annotation on `TypeDescriptor.GetProperties`. Unfortunately we had to use `DynamicallyAccessedMemberTypes.All` annotation on the `GetProperties` (since under the hood it creates a real type descriptor which needs to know pretty much everything about the type).
To fix the first warning (IL2077) change the `TypeStoreItem.DynamicallyAccessedTypes` to `DynamicallyAccessedMemberTypes.All`.

To fix the second one (IL2026) it will be necessary to add `RequiresUnreferencedCode` attribute onto the calling method (`CreatePropertyStoreItems`). That will in turn force you to add that attribute to all its callers and so on. Fortunately it seems that all the caller up one level already have the RUC attribute on them, so the change should be relatively simple.

The last warning (IL2072) will need to be suppressed by adding the `RequiresUnreferencedCode` attribute onto the `GetExplicitAttribute`. Reasoning is the same as the RUC On TypeDescriptor.GetProperties.

Reply:
Thanks @vitek-karas. I was avoiding changing the annotation on the `_type` field for some reason. I should have a PR in shortly.